ntoskrnl/mm/ARM3/sysldr.c

```diff
@@ -268,28 +256,168 @@ MiLocateExportName(IN PVOID DllBase,
     }
 
     /* Check if we couldn't find it */
-    if (High < Low) return NULL;
+    if (High < Low)
+        return -1;
 
     /* Otherwise, this is the ordinal */
-    Ordinal = OrdinalTable[Mid];
+    return OrdinalTable[Mid];
+}
+
+/**
+ * @brief
+ * ReactOS-only helper routine for RtlFindExportedRoutineByName(),
+ * that gives a finer granularity regarding the returned error codes.
+ *
+ * @param[in]   DllBase
+ * The base address of the loaded image.
+ *
+ * @param[in]   ExportName
+ * The name of the export, given as an ANSI NULL-terminated string.
+ *
+ * @param[out]  EntryPoint
+ * The address of the named exported routine, or NULL if not found.
+ *
+ * @param[in]   CheckIfForwarder
+ * If set to TRUE, check whether the found export is a forwarder. If so,
+ * the function fails with the STATUS_PROCEDURE_NOT_FOUND status code.
+ *
+ * @return
+ * A status code as follows:
+ * - STATUS_SUCCESS if the named exported routine could be found;
+ * - STATUS_INVALID_PARAMETER if the image is invalid or does not
+ *   contain an Export Directory;
+ * - STATUS_ORDINAL_NOT_FOUND if the export could not be found;
+ * - STATUS_PROCEDURE_NOT_FOUND if the export is a forwarder.
+ *
+ * @remark
+ * The function does not handle forwarders, and will fail with
+ * STATUS_PROCEDURE_NOT_FOUND if the export is as such.
+ *
+ * @note
+ * See RtlFindExportedRoutineByName() below for more remarks.
+ * Used by psmgr.c LookupEntryPoint() as well.
+ **/
+NTSTATUS
+NTAPI
+RtlpFindExportedRoutineByName(
+    _In_ PVOID DllBase,
+    _In_ PCSTR ExportName,
+    _Out_ PVOID* Function,
+    _In_opt_ BOOLEAN CheckIfForwarder)
+{
+    PIMAGE_EXPORT_DIRECTORY ExportDirectory;
+    PULONG NameTable;
+    PUSHORT OrdinalTable;
+    ULONG ExportSize;
+    USHORT Ordinal;
+    PULONG ExportTable;
+    PVOID FunctionAddress;
+
+    PAGED_CODE();
+
+    /* Get the export directory */
+    ExportDirectory = RtlImageDirectoryEntryToData(DllBase,
+                                                   TRUE,
+                                                   IMAGE_DIRECTORY_ENTRY_EXPORT,
+                                                   &ExportSize);
+    if (!ExportDirectory)
+        return STATUS_INVALID_PARAMETER;
+
+    /* Setup name tables */
+    NameTable = (PULONG)((ULONG_PTR)DllBase +
+                         ExportDirectory->AddressOfNames);
+    OrdinalTable = (PUSHORT)((ULONG_PTR)DllBase +
+                             ExportDirectory->AddressOfNameOrdinals);
+
+    /* Get the ordinal */
+    Ordinal = NameToOrdinal(ExportName,
+                            DllBase,
+                            ExportDirectory->NumberOfNames,
+                            NameTable,
+                            OrdinalTable);
+
+    /* Check if we couldn't find it */
+    if (Ordinal == -1)
+        return STATUS_ORDINAL_NOT_FOUND;
+
+    /* Validate the ordinal */
+    if (Ordinal >= ExportDirectory->NumberOfFunctions)
```

Comment:
Do you need to take into account `IMAGE_EXPORT_DIRECTORY.Base` here?

![image](https://user-images.githubusercontent.com/610685/204516525-59b4ed41-3c05-40ab-8d63-d0be491c4a50.png)

Reply:
No; this PR code is functionally equivalent to the old one we had (and what appears to be even in Win10).